zanata-war/src/main/java/org/zanata/adapter/GettextAdapter.java

```diff
@@ -0,0 +1,171 @@
+
+        if (sourceLocale == null) {
+            throw new IllegalArgumentException("Source locale cannot be null");
+        }
+
+        if (fileUri == null) {
+            throw new IllegalArgumentException("Document URI cannot be null");
+        }
+
+        PoReader2 reader = new PoReader2();
+
+        BufferedInputStream inputStream = readStream(fileUri);
+        Resource doc = reader.extractTemplate(new InputSource(inputStream),
+                        sourceLocale, "");
+        try {
+            inputStream.close();
+        } catch (IOException e) {
+        }
+        return doc;
+    }
+
+    @Override
+    public TranslationsResource parseTranslationFile(URI fileUri,
+        String localeId, Optional<String> params)
+        throws FileFormatAdapterException, IllegalArgumentException {
+
+        if (StringUtils.isEmpty(localeId)) {
+            throw new IllegalArgumentException(
+                "locale id string cannot be null or empty");
+        }
+
+        PoReader2 reader = new PoReader2();
+
+        BufferedInputStream inputStream = readStream(fileUri);
+        TranslationsResource resource =
+                reader.extractTarget(new InputSource(inputStream));
+
+        try {
+            inputStream.close();
+        } catch (IOException e) {
+        }
+
+        return resource;
+    }
+
+    @Override
+    public void writeTranslatedFile(OutputStream output, URI originalFile,
+        Resource resource, TranslationsResource translationsResource,
+        String locale, Optional<String> params)
+        throws FileFormatAdapterException, IllegalArgumentException {
+
+        PoWriter2 writer = new PoWriter2(true, true);
+        try {
+            writer.writePo(output, Charsets.UTF_8.name(), resource,
+                    translationsResource);
+        } catch (IOException e) {
+            throw new FileFormatAdapterException(
+                "Unable to generate translated file", e);
+        }
+    }
+
+    private BufferedInputStream readStream(URI fileUri) throws FileFormatAdapterException,
+        IllegalArgumentException {
+        URL url = null;
+
+        try {
+            url = fileUri.toURL();
+            return new BufferedInputStream(url.openStream());
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(
+                "Could not open the URI. The URI must be absolute: "
+                    + ((url == null) ? "URL is null" : url.toString()),
+                e);
+        } catch (MalformedURLException e) {
+            throw new FileFormatAdapterException(
+                "Could not open the URI. The URI may be malformed: "
+                    + ((url == null) ? "URL is null" : url.toString()),
+                e);
+        } catch (IOException e) {
+            throw new FileFormatAdapterException(
+                "Could not open the URL. The URL is OK but the input stream could not be opened.\n"
+                    + e.getMessage(), e);
```

Comment:
Can these exception messages use format strings with placeholders instead? I forget whether that is standard for exceptions.